dp-hello-world-event/event/consumer.go

```diff
@@ -0,0 +1,104 @@
+package event
+
+import (
+	"context"
+	"errors"
+
+	"github.com/ONSdigital/dp-hello-world-event/schema"
+	kafka "github.com/ONSdigital/dp-kafka/v2"
+	"github.com/ONSdigital/log.go/log"
+)
+
+//go:generate moq -out mock/handler.go -pkg mock . Handler
+
+// MessageConsumer provides a generic interface for consuming []byte messages
+type MessageConsumer interface {
+	Channels() *kafka.ConsumerGroupChannels
+}
+
+// Handler represents a handler for processing a single event.
+type Handler interface {
+	Handle(ctx context.Context, helloCalled *HelloCalled) error
+}
+
+// Consumer consumes event messages.
+type Consumer struct {
+	closing chan bool
+	closed  chan bool
+}
+
+// NewConsumer returns a new consumer instance.
+func NewConsumer() *Consumer {
+	return &Consumer{
+		closing: make(chan bool),
+		closed:  make(chan bool),
+	}
+}
+
+// Consume converts messages to event instances, and pass the event to the provided handler.
+func (consumer *Consumer) Consume(ctx context.Context, messageConsumer MessageConsumer, handler Handler) {
+
+	go func() {
+		defer close(consumer.closed)
+
+		for {
+			select {
+			case message := <-messageConsumer.Channels().Upstream:
+				messageCtx := context.Background()
+				processMessage(messageCtx, message, handler)
+				message.Release()
+			case <-consumer.closing:
+				log.Event(ctx, "closing event consumer loop", log.INFO)
```

Comment:
We have a 'StopListeningToConsumer' function that allows use to still have some time while a message thats in flight is being processed. I think its why we have 'closing' and 'closed' channels. StopListeningToConsumer should potentially be called in here? Or it might be called wherever we're sending a message onto the 'closing' channel, I'm not sure